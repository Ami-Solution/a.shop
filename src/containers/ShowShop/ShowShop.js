import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { toast } from 'react-toastify';

/*
  Components
 */
import { ShopRecap, Button, LoaderScreen } from '../../components';
import BetaModal from '../Wrapper/BetaModal';

/*
  Translate module
 */
import tr from '../../translate';

/*
  Redux
 */
import {
  addDeleteShopTransaction as addDeleteShopTransactionAction,
  removeShop as removeShopAction,
  endTransaction as endTransactionAction,
} from '../../actions/shop';
import { fetchAll as fetchAllAction } from '../../actions/map';

/*
  Helpers
 */
import {
  deleteShop as deleteShopHelper,
  getTransactionStatus,
} from '../../helpers';

/**
 * ShowShop container
 * @extends PureComponent
 */
export class ShowShop extends PureComponent {
  static propTypes = {
    shop: PropTypes.shape({
      name: PropTypes.string.isRequired,
      cat: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      opening: PropTypes.string.isRequired,
      lat: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
      lng: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
    }).isRequired,
    addDeleteShopTransaction: PropTypes.func.isRequired,
    isTransactionPending: PropTypes.bool.isRequired,
    deleteContractShop: PropTypes.func.isRequired,
    removeShopFromStore: PropTypes.func.isRequired,
    transactionHash: PropTypes.string.isRequired,
    endTransaction: PropTypes.func.isRequired,
    fetchAll: PropTypes.func.isRequired,
    centerPosition: PropTypes.shape({}).isRequired,
  };

  state = {
    isLoading: false,
    showModal: false,
  };
  componentWillMount() {
    const { isTransactionPending } = this.props;

    if (isTransactionPending) {
      this.interval = this.checkTransaction();
    }
  }

  componentWillUnmount() {
    if (this.interval) {
      clearInterval(this.interval);
    }
  }

  getView = () => {
    const { isTransactionPending } = this.props;

    if (isTransactionPending)
      return <div>{tr('show_shop.transaction_pending')}</div>;
    return (
      <Button onClick={this.deleteShop}>{tr('show_shop.delete_button')}</Button>
    );
  };

  showLoader = () => this.setState({ isLoading: true });
  HideLoader = () => this.setState({ isLoading: false });

  endCheckTransaction = () => {
    const { endTransaction } = this.props;
    endTransaction();
    clearInterval(this.interval);
  };

  checkTransaction = () => {
    const {
      transactionHash,
      removeShopFromStore,
      fetchAll,
      centerPosition,
    } = this.props;
    this.interval = setInterval(async () => {
      const status = await getTransactionStatus(transactionHash);
      if (status === 'success') {
        removeShopFromStore();
        fetchAll(centerPosition);
        this.endCheckTransaction();
      } else if (status === 'error') {
        this.endCheckTransaction();
        toast.error(tr('errors.transaction.throw'));
      }
    }, 3000);
  };

  deleteShop = async () => {
    const { deleteContractShop, addDeleteShopTransaction } = this.props;

    this.showLoader();
    try {
      const transaction = await deleteContractShop();
      addDeleteShopTransaction(transaction.transactionHash);
      this.checkTransaction();
      this.HideLoader();
    } catch (e) {
      // toast.error(tr('errors.transaction.metamask_reject'));
      // this.HideLoader();
      this.setState({ showModal: true });
    }
  };

  closeModal = () => {
    this.setState({ showModal: false });
  };

  render = () => {
    const { shop } = this.props;
    const { isLoading, showModal } = this.state;

    return isLoading ? (
      <LoaderScreen
        title={tr('show_shop.loader_title')}
        message={tr('show_shop.loader_delete_message')}
        isTransaction
      />
    ) : (
      <Fragment>
        {showModal && <BetaModal close={this.closeModal} send />}
        <ShopRecap {...shop} />
        {this.getView()}
      </Fragment>
    );
  };
}

const mapStateToProps = ({ shop, map }) => ({
  shop: shop.shop,
  isTransactionPending: !!shop.transactionHash,
  transactionHash: shop.transactionHash || '',
  centerPosition: map.centerPosition,
});

const mapDispatchToProps = dispatch => ({
  deleteContractShop: deleteShopHelper,
  removeShopFromStore: bindActionCreators(removeShopAction, dispatch),
  addDeleteShopTransaction: bindActionCreators(
    addDeleteShopTransactionAction,
    dispatch,
  ),
  endTransaction: bindActionCreators(endTransactionAction, dispatch),
  fetchAll: bindActionCreators(fetchAllAction, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(ShowShop);
