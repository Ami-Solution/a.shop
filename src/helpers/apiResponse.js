import tr from '../translate';

export const getErrorMessage = (errors, { response }) => {
  const { status } = response;
  if (status === 429) {
    return tr('tooManyRequests');
  } else if (Array.isArray(errors.message)) {
    return errors.message.join('; ');
  } else if (errors.message) {
    return errors.message;
  }
  return response.statusText;
};
