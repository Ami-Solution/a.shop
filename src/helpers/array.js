export const uniqueArray = (arr, objKey) => {
  const keyToPos = {};

  arr.forEach((e, idx) => { keyToPos[e[objKey]] = idx; });
  return arr.filter((e, idx) => keyToPos[e[objKey]] === idx);
};
