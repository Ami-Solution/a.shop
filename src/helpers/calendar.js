export const times = [
  { key: 'a', val: '00:00' },
  { key: 'b', val: '00:30' },
  { key: 'c', val: '01:00' },
  { key: 'd', val: '01:30' },
  { key: 'e', val: '02:00' },
  { key: 'f', val: '02:30' },
  { key: 'g', val: '03:00' },
  { key: 'h', val: '03:30' },
  { key: 'i', val: '04:00' },
  { key: 'j', val: '04:30' },
  { key: 'k', val: '05:00' },
  { key: 'l', val: '05:30' },
  { key: 'm', val: '06:00' },
  { key: 'n', val: '06:30' },
  { key: 'o', val: '07:00' },
  { key: 'p', val: '07:30' },
  { key: 'q', val: '08:00' },
  { key: 'r', val: '08:30' },
  { key: 's', val: '09:00' },
  { key: 't', val: '09:30' },
  { key: 'u', val: '10:00' },
  { key: 'v', val: '10:30' },
  { key: 'w', val: '11:00' },
  { key: 'x', val: '11:30' },
  { key: 'y', val: '12:00' },
  { key: 'z', val: '12:30' },
  { key: 'A', val: '13:00' },
  { key: 'B', val: '13:30' },
  { key: 'C', val: '14:00' },
  { key: 'D', val: '14:30' },
  { key: 'E', val: '15:00' },
  { key: 'F', val: '15:30' },
  { key: 'G', val: '16:00' },
  { key: 'H', val: '16:30' },
  { key: 'I', val: '17:00' },
  { key: 'J', val: '17:30' },
  { key: 'K', val: '18:00' },
  { key: 'L', val: '18:30' },
  { key: 'M', val: '19:00' },
  { key: 'N', val: '19:30' },
  { key: 'O', val: '20:00' },
  { key: 'P', val: '20:30' },
  { key: 'Q', val: '21:00' },
  { key: 'R', val: '21:30' },
  { key: 'S', val: '22:00' },
  { key: 'T', val: '22:30' },
  { key: 'U', val: '23:00' },
  { key: 'V', val: '23:30' }
];

export const hours = times.map(e => e.val);

export const convertDay = ({ open, openAt, closeAt }) => {
  if (!open) {
    return '0';
  }
  const charOpen = times.filter(e => e.val === openAt)[0].key;
  const charClose = times.filter(e => e.val === closeAt)[0].key;
  return charOpen + charClose;
};
