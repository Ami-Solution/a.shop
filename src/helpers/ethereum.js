import Web3 from 'web3';
import DthContract from 'dethercontract/contracts/DetherToken.json';

const web3js = new Web3(window.web3.currentProvider);

/**
 * know if user as web3 injected
 * @returns {Bool}
 */
export const isWeb3 = () => {
  if (typeof window.web3 !== 'undefined') {
    return true;
  } else {
    return false;
  }
};

/**
 * return balance of ETH and DTH of the web3[0]
 * @returns {Object}
 */
export const getBalance = () =>
  new Promise(async (res, rej) => {
    let balances = {}
    // const dthInstance = dthContract.at('0x84f073b99FB93812DC701874956226319eD17b7c');
    const address =  (await web3js.eth.getAccounts())[0];

    balances.dth = 1000;
    try {
      web3js.eth.getBalance(address).then((result, error) => {
        if(!error){
          balances.eth = result;
          res(balances);
        } else {
          console.log('error transaction', error);
          rej({error: 'provider call failed'});
        }
      });
    } catch (e) {
      rej(e)
    }
  });
