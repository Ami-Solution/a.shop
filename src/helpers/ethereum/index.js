export { default as addShop } from './addShop';
export { default as deleteShop } from './deleteShop';
export { default as getBalance } from './getBalance';
export { default as getShop } from './getShop';
export { default as isWeb3 } from './isWeb3';
