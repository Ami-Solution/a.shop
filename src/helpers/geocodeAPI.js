import config from '../constants/config';

const GOOGLE_API = 'https://maps.google.com/maps/api/geocode/json';

const filterPostalCode = e => e.types.includes('postal_code') && !!e.long_name;

const GeocodeAPI = {
  getPostalCodeFromAddressComponents(addressComponents) {
    const elemFind = addressComponents.filter(filterPostalCode);
    if (elemFind && elemFind.length) {
      return elemFind[0].long_name;
    }
    return null;
  },

  getCountryIdFromAddressComponents(addressComponents) {
    const elemFind = addressComponents.filter(e => e.types.includes('country') && !!e.short_name);
    if (elemFind && elemFind.length) {
      return elemFind[0].short_name;
    }
    return null;
  },

  /**
   * @param {Object} lat and lng
   * @returns {Promise} the postal code of the lat and lng
   */
  async postalCode({ lat, lng }) {
    const results = await this.reverseGeocode({ lat, lng });

    for (let i = 0; i < results.length; i += 1) {
      const elemFind = this.getPostalCodeFromAddressComponents(results[i].address_components);
      if (elemFind) {
        return elemFind;
      }
    }
    return null;
  },

  /**
   * @param {Object} lat and lng
   * @returns {Promise}
   */
  async reverseGeocode({ lat, lng }) {
    if (!lat || !lng) {
      return Promise.reject(new Error('Provided coordinates are invalid'));
    }

    const latLng = `${lat},${lng}`;
    const url = `${GOOGLE_API}?key=${config.googleMapKey}&latlng=${encodeURI(latLng)}`;
    const { results } = await this.handleUrl(url);
    if (!results || results.length <= 0) {
      return Promise.reject(new Error('Enable to reverse lat lng'));
    }

    return results;
  },

  async handleUrl(url) {
    const response = await fetch(url).catch(() => Promise.reject(new Error('Error fetching data')));

    const json = await response
      .json()
      .catch(() => Promise.reject(new Error('Error parsing server response')));

    if (json.status === 'OK') {
      return json;
    }
    return Promise.reject(new Error(`Server returned status code ${json.status}`));
  }
};

export default GeocodeAPI;
