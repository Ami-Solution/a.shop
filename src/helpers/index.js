export get from './get';
export last from './last';
