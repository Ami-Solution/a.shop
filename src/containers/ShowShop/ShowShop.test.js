/* global describe it expect */
import React from 'react';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';
import TestUtils from 'react-dom/test-utils';
import { BrowserRouter as Router } from 'react-router-dom';

import { GeocodeAPI } from '../../helpers';
import fakeStore from '../../constants/test/fakeStore';
import { ShowShop } from './ShowShop';
import jsdom from 'jsdom'
const { JSDOM } = jsdom;


/*** Set up test environment to run in a browser-like environment ***/
const { document } = (new JSDOM('<!doctype html><html><body></body></html>')).window;

global.document = document;
global.window = global.document.defaultView

configure({ adapter: new Adapter() });

const store = fakeStore({});

GeocodeAPI.positionToAddress = async () => 'Paris 2';
GeocodeAPI.postalCode = async () => '75002';

describe('Container ShowShop', () => {
  it('should be render without crash', () => {
    const component = TestUtils.renderIntoDocument(
        <ShowShop
          shop={{
            name: 'test',
            cat: 'more test',
            description: 'more more test',
            opening: '0000000',
            lat: '48.8628',
            lng: '2.3292'
          }}
          addShopToStore={() => {}}
          addShopToContract={() => {}}
          isTransactionPending={false}
          deleteContractShop={() => {}}
        />
    );
    expect(component).not.toBe(null);
  });
});
