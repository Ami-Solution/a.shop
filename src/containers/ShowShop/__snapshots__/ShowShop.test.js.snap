// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowShop::mount 1`] = `undefined`;

exports[`ShowShop::shallow 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShowShop
    addDeleteShopTransaction={[Function]}
    centerPosition={
        Object {
            "lat": 30,
            "lng": 3,
          }
    }
    deleteContractShop={[Function]}
    endTransaction={[Function]}
    fetchAll={[Function]}
    isTransactionPending={false}
    removeShopFromStore={[Function]}
    shop={
        Object {
            "address": "40 rue Citadelle Paris",
            "cat": "cool",
            "countryId": "fr",
            "description": "Super Magasin",
            "lat": "30",
            "lng": "3",
            "name": "Magasin",
            "opening": "0000000",
            "postalCode": "75010",
          }
    }
    transactionHash="0xwefwerfer"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        false,
        <ShopRecap
          address="40 rue Citadelle Paris"
          cat="cool"
          countryId="fr"
          description="Super Magasin"
          lat="30"
          licencePrice={null}
          lng="3"
          name="Magasin"
          opening="0000000"
          postalCode="75010"
/>,
        <Button
          disabled={false}
          fullWidth={false}
          isSmall={false}
          onClick={[Function]}
          theme="light"
          width={null}
>
          Delete Shop
</Button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "address": "40 rue Citadelle Paris",
          "cat": "cool",
          "countryId": "fr",
          "description": "Super Magasin",
          "lat": "30",
          "licencePrice": null,
          "lng": "3",
          "name": "Magasin",
          "opening": "0000000",
          "postalCode": "75010",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Delete Shop",
          "disabled": false,
          "fullWidth": false,
          "isSmall": false,
          "onClick": [Function],
          "theme": "light",
          "width": null,
        },
        "ref": null,
        "rendered": "Delete Shop",
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          false,
          <ShopRecap
            address="40 rue Citadelle Paris"
            cat="cool"
            countryId="fr"
            description="Super Magasin"
            lat="30"
            licencePrice={null}
            lng="3"
            name="Magasin"
            opening="0000000"
            postalCode="75010"
/>,
          <Button
            disabled={false}
            fullWidth={false}
            isSmall={false}
            onClick={[Function]}
            theme="light"
            width={null}
>
            Delete Shop
</Button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "address": "40 rue Citadelle Paris",
            "cat": "cool",
            "countryId": "fr",
            "description": "Super Magasin",
            "lat": "30",
            "licencePrice": null,
            "lng": "3",
            "name": "Magasin",
            "opening": "0000000",
            "postalCode": "75010",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Delete Shop",
            "disabled": false,
            "fullWidth": false,
            "isSmall": false,
            "onClick": [Function],
            "theme": "light",
            "width": null,
          },
          "ref": null,
          "rendered": "Delete Shop",
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
